src/mdconvert.1
.\" SPDX-FileCopyrightText: 2004-2022 Oswald Buddenhagen <ossi@users.sf.net>
.\" SPDX-License-Identifier: GPL-2.0-or-later
.\"
.\" mdconvert - Maildir mailbox UID storage scheme converter
.
.TH mdconvert 1 "2004 Mar 27"
.
.SH NAME
mdconvert - Maildir mailbox UID storage scheme converter
.
.SH SYNOPSIS
\fBmdconvert\fR [\fIoptions\fR ...] \fImailbox\fR ...
.
.SH DESCRIPTION
\fBmdconvert\fR converts Maildir mailboxes between the two UID storage schemes
supported by \fBmbsync\fR. See \fBmbsync\fR's manual page for details on these
schemes.
.
.SH OPTIONS
.TP
\fB-a\fR, \fB--alt\fR
Convert to the \fBalternative\fR (Berkeley DB based) UID storage scheme.
.TP
\fB-n\fR, \fB--native\fR
Convert to the \fBnative\fR (file name based) UID storage scheme.
This is the default.
.TP
\fB-h\fR, \fB--help\fR
Displays a summary of command line options.
.TP
\fB-v\fR, \fB--version\fR
Displays version information.
.
.SH SEE ALSO
mbsync(1)
.
.SH AUTHOR
Written and maintained by Oswald Buddenhagen <ossi@users.sf.net>.


NEWS
[1.5.0]

Changed default config & state locations to follow the XDG basedir spec.
The old locations remain supported.

The reference point for relative local paths in the configuration file
is now the file's containing directory.

Placeholders will be now created for messages exceeding MaxSize even if
they are flagged on the source side.

Renamed the ReNew/--renew/-N options to Upgrade/--upgrade/-u
and Delete/--delete/-d to Gone/--gone/-g.

MaxMessages and MaxSize can be used together now.

The unfiltered list of mailboxes in each Store can be printed now.

A proper summary is now printed prior to exiting.

Added new sync operation 'Old'.

[1.4.0]

The 'isync' compatibility wrapper was removed.

Added support for disabling TLS v1.3 - adjust SSLVersions if you set it.
Removed support for obsolete/insecure SSL v3.

The IMAP '$Forwarded' / Maildir 'P' (passed) flag is supported now.

Support for configuring a TLS cipher string was added.

IMAP mailbox subscriptions are supported now.

The IMAP user query can be scripted now.

Added built-in support for macOS Keychain.

Messages excluded by MaxSize will now result in placeholders.

The use of Master/Slave terminology has been deprecated.

[1.3.0]

Network timeout handling has been added.

Support for proper Maildir++ and a Maildir sub-folder naming style
without extra dots have been added.

Support for TLS client certificates was added.

Support for recovering from baseless UID validity changes was added.

The get-cert script was renamed to mbsync-get-cert.

[1.2.0]

The 'isync' compatibility wrapper is now deprecated.

An IMAP Path/NAMESPACE rooted in INBOX won't be handled specially any more.
This means that some Patterns may need adjustment.

The default output is a lot less verbose now.
The meanings of the -V and -D options changed significantly.

The SSL/TLS configuration has been re-designed.
SSL is now explicitly enabled or disabled - "use SSL if available" is gone.
Notice: Tunnels are assumed to be secure and thus default to no SSL.

Support for SASL (flexible authentication) has been added.

Support for Windows file systems has been added.

Support for compressed data transfer has been added.

Folder deletions can be propagated now.

[1.1.0]

Support for hierarchical mailboxes in Patterns.

Full support for IMAP pipelining (streaming, parallelization) added.
This is considerably faster especially with high-latency networks.

Faster and hopefully more reliable support for IMAP servers without the
UIDPLUS extension (e.g., M$ Exchange).

More automatic handling of SSL certificates.

IPv6 support.

IMAP password query can be scripted.

Message arrival dates can be propagated.

Data safety in case of system crashes was improved.

MaxMessages was made vastly more useful.

[1.0.0]

Essentially a rewrite. Synchronization state storage concept, configuration
and command line changed entirely.
But you needn't to worry about the upgrade, as a fully automated migration
path is provided, even for users of isync 0.7 and below.
Still, you should re-read the manual to be able to take full advantage of the
new features:

The supported mailbox types can be freely paired.
A possible application of this is using a local IMAP server to access
mailboxes that are not natively supported yet.

Message deletions (expunges) are now propagated both ways, so there is no need
for using mutt with maildir_trash any more.

Additional trash options added.

`OneToOne' replaced by something more flexible.

Partial support for IMAP pipelining (streaming, parallelization) added.
Makes flag change propagation much faster - this affects every message that
becomes Seen/Read.

[0.9]

Added Tunnel directive to allow the user to specify a shell command to run
to set up an IMAP connection in place of a TCP socket (eg., to run over
an SSH session).

Added PREAUTH support (useful mostly in conjunction with Tunnel).

Messages marked deleted are not uploaded when we are going to expunge.

Locally generated messages are not re-fetched after uploading even if the
UIDPLUS extension is not supported by the server.

Added `OneToOne' configuration option: ignore any Mailbox specifications
and instead pick up all mailboxes from the local MailDir and remote Folder
and map them 1:1 onto each other according to their names.

-C now creates both local and remote boxes; -L and -R create only local/remote.

--quiet is now really quiet.

[0.8]

!!! IMPORTANT !!!

In order to fix the problem where messages copied from one mailbox to
another were not uploaded to the new mailbox, the way Isync stores the UID
for each message needed to be changed.  As a result, you _MUST_ delete all
the messages in the local maildir box before using this version.  Otherwise
it will upload every message to the server thinking its a new mail.

[0.7]

Added `MaxMessages' configuration option to allow tracking of only the most
recently added message in the local mailbox.

Added --create (-C) command line option to force creation of the local
maildir-style mailbox if it doesn't already exist.

[0.6]

Added `Delete' configuration option to correspond to the -d command line
option.

Added -a (--all) command line option to synchronize all mailboxes.

[0.5]

Updated SSL support.

Added CRAM authentication support.

Added MailDir configuration option to specify the default location of local
mailboxes when relative paths are used.

Added support for uploading local messages to the IMAP server.

Added CopyDeletedTo configuration option to cause isync to move deleted
messages to a particular mailbox on the server when they are expunged.

[0.4]

Added MaxSize configuration option to limit downloading of new messages from
the server to less than some threshold.

More robust --fast option works without using \Recent flags, so the previous
problem with multiple accesses killing these flags is no longer a problem.

RFC2060 obsoleted RFC822.PEEK, use BODY.PEEK[] instead which does the same
job.


src/mdconvert.c
// SPDX-FileCopyrightText: 2004-2022 Oswald Buddenhagen <ossi@users.sf.net>
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 * mdconvert - Maildir UID scheme converter
 */

#include <autodefs.h>

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <stdarg.h>
#include <dirent.h>
#include <limits.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>

#include <db.h>

#define EXE "mdconvert"

#if __GNUC__ > 2 || (__GNUC__ == 2 && __GNUC_MINOR__ > 4)
# define ATTR_NORETURN __attribute__((noreturn))
# define ATTR_PRINTFLIKE(fmt,var) __attribute__((format(printf,fmt,var)))
#else
# define ATTR_NORETURN
# define ATTR_PRINTFLIKE(fmt,var)
#endif

static void ATTR_NORETURN
oob( void )
{
	fputs( "Fatal: buffer too small. Please report a bug.\n", stderr );
	abort();
}

static void ATTR_PRINTFLIKE(1, 2)
sys_error( const char *msg, ... )
{
	va_list va;
	char buf[1024];

	va_start( va, msg );
	if ((unsigned)vsnprintf( buf, sizeof(buf), msg, va ) >= sizeof(buf))
		oob();
	va_end( va );
	perror( buf );
}

static int ATTR_PRINTFLIKE(3, 4)
nfsnprintf( char *buf, int blen, const char *fmt, ... )
{
	int ret;
	va_list va;

	va_start( va, fmt );
	if (blen <= 0 || (unsigned)(ret = vsnprintf( buf, blen, fmt, va )) >= (unsigned)blen)
		oob();
	va_end( va );
	return ret;
}

static const char *subdirs[] = { "cur", "new" };
static struct flock lck;
static DBT key, value;

static int
convert( const char *box, int altmap )
{
	DB *db;
	DIR *d;
	struct dirent *e;
	const char *u, *ru;
	char *p, *s, *dpath, *spath, *dbpath;
	int i, n, ret, sfd, dfd, bl, ml, uv[2], uid;
	struct stat st;
	char buf[_POSIX_PATH_MAX], buf2[_POSIX_PATH_MAX];
	char umpath[_POSIX_PATH_MAX], uvpath[_POSIX_PATH_MAX], tdpath[_POSIX_PATH_MAX];

	if (stat( box, &st ) || !S_ISDIR(st.st_mode)) {
		fprintf( stderr, "'%s' is no Maildir mailbox.\n", box );
		return 1;
	}

	nfsnprintf( umpath, sizeof(umpath), "%s/.isyncuidmap.db", box );
	nfsnprintf( uvpath, sizeof(uvpath), "%s/.uidvalidity", box );
	if (altmap)
		dpath = umpath, spath = uvpath, dbpath = tdpath;
	else
		spath = umpath, dpath = uvpath, dbpath = umpath;
	nfsnprintf( tdpath, sizeof(tdpath), "%s.tmp", dpath );
	if ((sfd = open( spath, O_RDWR )) < 0) {
		if (errno != ENOENT)
			sys_error( "Cannot open %s", spath );
		return 1;
	}
	if (fcntl( sfd, F_SETLKW, &lck )) {
		sys_error( "Cannot lock %s", spath );
		goto sbork;
	}
	if ((dfd = open( tdpath, O_RDWR|O_CREAT, 0600 )) < 0) {
		sys_error( "Cannot create %s", tdpath );
		goto sbork;
	}
	if (db_create( &db, NULL, 0 )) {
		fputs( "Error: db_create() failed\n", stderr );
		goto tbork;
	}
	if ((ret = (db->open)( db, NULL, dbpath, NULL, DB_HASH, altmap ? DB_CREATE|DB_TRUNCATE : 0, 0 ))) {
		db->err( db, ret, "Error: db->open(%s)", dbpath );
	  dbork:
		db->close( db, 0 );
	  tbork:
		unlink( tdpath );
		close( dfd );
	  sbork:
		close( sfd );
		return 1;
	}
	key.data = (void *)"UIDVALIDITY";
	key.size = 11;
	if (altmap) {
		if ((n = read( sfd, buf, sizeof(buf) - 1 )) <= 0 ||
		    (buf[n] = 0, sscanf( buf, "%d\n%d", &uv[0], &uv[1] ) != 2))
		{
			fprintf( stderr, "Error: cannot read UIDVALIDITY of '%s'.\n", box );
			goto dbork;
		}
		value.data = uv;
		value.size = sizeof(uv);
		if ((ret = db->put( db, NULL, &key, &value, 0 ))) {
			db->err( db, ret, "Error: cannot write UIDVALIDITY for '%s'", box );
			goto dbork;
		}
	} else {
		if ((ret = db->get( db, NULL, &key, &value, 0 ))) {
			db->err( db, ret, "Error: cannot read UIDVALIDITY of '%s'", box );
			goto dbork;
		}
		n = sprintf( buf, "%d\n%d\n", ((int *)value.data)[0], ((int *)value.data)[1] );
		if (write( dfd, buf, n ) != n) {
			fprintf( stderr, "Error: cannot write UIDVALIDITY for '%s'.\n", box );
			goto dbork;
		}
	}

  again:
	for (i = 0; i < 2; i++) {
		bl = nfsnprintf( buf, sizeof(buf), "%s/%s/", box, subdirs[i] );
		if (!(d = opendir( buf ))) {
			sys_error( "Cannot list %s", buf );
			goto dbork;
		}
		while ((e = readdir( d ))) {
			if (*e->d_name == '.')
				continue;
			nfsnprintf( buf + bl, sizeof(buf) - bl, "%s", e->d_name );
			memcpy( buf2, buf, bl );
			p = strstr( e->d_name, ",U=" );
			if (p)
				for (u = p, ru = p + 3; isdigit( (unsigned char)*ru ); ru++);
			else
				u = ru = strchr( e->d_name, ':' );
			if (u)
				ml = u - e->d_name;
			else
				ru = "", ml = sizeof(buf);
			if (altmap) {
				if (!p)
					continue;
				key.data = e->d_name;
				key.size = (size_t)(strchr( e->d_name, ',' ) - e->d_name);
				uid = atoi( p + 3 );
				value.data = &uid;
				value.size = sizeof(uid);
				if ((ret = db->put( db, NULL, &key, &value, 0 ))) {
					db->err( db, ret, "Error: cannot write UID for '%s'", box );
					goto ebork;
				}
				nfsnprintf( buf2 + bl, sizeof(buf2) - bl, "%.*s%s", ml, e->d_name, ru );
			} else {
				s = strpbrk( e->d_name, ",:" );
				key.data = e->d_name;
				key.size = s ? (size_t)(s - e->d_name) : strlen( e->d_name );
				if ((ret = db->get( db, NULL, &key, &value, 0 ))) {
					if (ret != DB_NOTFOUND) {
						db->err( db, ret, "Error: cannot read UID for '%s'", box );
						goto ebork;
					}
					continue;
				}
				uid = *(int *)value.data;
				nfsnprintf( buf2 + bl, sizeof(buf2) - bl, "%.*s,U=%d%s", ml, e->d_name, uid, ru );
			}
			if (rename( buf, buf2 )) {
				if (errno == ENOENT) {
					closedir( d );
					goto again;
				}
				sys_error( "Cannot rename %s to %s", buf, buf2 );
			  ebork:
				closedir( d );
				goto dbork;
			}

		}
		closedir( d );
	}

	db->close( db, 0 );
	close( dfd );
	if (rename( tdpath, dpath )) {
		sys_error( "Cannot rename %s to %s", tdpath, dpath );
		close( sfd );
		return 1;
	}
	if (unlink( spath ))
		sys_error( "Cannot remove %s", spath );
	close( sfd );
	return 0;
}

int
main( int argc, char **argv )
{
	int oint, ret, altmap = 0;

	for (oint = 1; oint < argc; oint++) {
		if (!strcmp( argv[oint], "-h" ) || !strcmp( argv[oint], "--help" )) {
			puts(
"Usage: " EXE " [-a] mailbox...\n"
"  -a, --alt      convert to alternative (DB based) UID scheme\n"
"  -n, --native   convert to native (file name based) UID scheme (default)\n"
"  -h, --help     show this help message\n"
"  -v, --version  display version"
			);
			return 0;
		} else if (!strcmp( argv[oint], "-v" ) || !strcmp( argv[oint], "--version" )) {
			puts( EXE " " VERSION " - Maildir UID scheme converter" );
			return 0;
		} else if (!strcmp( argv[oint], "-a" ) || !strcmp( argv[oint], "--alt" )) {
			altmap = 1;
		} else if (!strcmp( argv[oint], "-n" ) || !strcmp( argv[oint], "--native" )) {
			altmap = 0;
		} else if (!strcmp( argv[oint], "--" )) {
			oint++;
			break;
		} else if (argv[oint][0] == '-') {
			fprintf( stderr, "Unrecognized option '%s'. Try " EXE " -h\n", argv[oint] );
			return 1;
		} else
			break;
	}
	if (oint == argc) {
		fprintf( stderr, "Mailbox specification missing. Try " EXE " -h\n" );
		return 1;
	}
#if SEEK_SET != 0
	lck.l_whence = SEEK_SET;
#endif
#if F_WRLCK != 0
	lck.l_type = F_WRLCK;
#endif
	ret = 0;
	for (; oint < argc; oint++)
		ret |= convert( argv[oint], altmap );
	return ret;
}

